### ISSUE TITLE
Error while syncing tables having identity column other than primary key

### ISSUE THREAD

**Issue by mangireesh** [Mimetis/Dotmim.Sync#337]
Hi,
I am new to this Syncing process. I am trying to sync MSSQL Server with SQLite Client database(Sync over HTTP). In my server some tables having identity columns(auto-increment) and primary key(unique id or GUID) both are different columns. While performing the sync operation its getting error on the identity column(ids_acc_gp). Is it possible to do the sync without changing the schema?

Dotmim.Sync.SyncException: SQLite Error 1: 'no such column: p.ids_acc_gp'. ---> Microsoft.Data.Sqlite.SqliteException: SQLite Error 1: 'no such column: p.ids_acc_gp'.
  at Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC (System.Int32 rc, SQLitePCL.sqlite3 db) [0x0006d] in <3f508591f2d746f08e36be64c74fd80d>:0 
  at Microsoft.Data.Sqlite.SqliteCommand+<PrepareAndEnumerateStatements>d__64.MoveNext () [0x000b0] in <3f508591f2d746f08e36be64c74fd80d>:0 
  at Microsoft.Data.Sqlite.SqliteCommand.Prepare () [0x00063] in <3f508591f2d746f08e36be64c74fd80d>:0 
  at Dotmim.Sync.SyncAdapter.PrepareCommandAsync (Dotmim.Sync.Builders.DbCommandType commandType, System.Data.Common.DbConnection connection, System.Data.Common.DbTransaction transaction, Dotmim.Sync.SyncFilter filter) [0x001a4] in <cc96665cb56c4641847da5ce877d6522>:0 
  at Dotmim.Sync.SyncAdapter.ApplyUpdateAsync (Dotmim.Sync.SyncRow row, System.Int64 lastTimestamp, System.Nullable`1[T] senderScopeId, System.Boolean forceWrite, System.Data.Common.DbConnection connection, System.Data.Common.DbTransaction transaction) [0x00095] in <cc96665cb56c4641847da5ce877d6522>:0 
  at Dotmim.Sync.SyncAdapter.ApplyChangesAsync (System.Guid localScopeId, System.Guid senderScopeId, Dotmim.Sync.SyncTable changesTable, System.Int64 lastTimestamp, System.Collections.Generic.List`1[T] conflicts, System.Data.Common.DbConnection connection, System.Data.Common.DbTransaction transaction) [0x000ff] in <cc96665cb56c4641847da5ce877d6522>:0 
  at Dotmim.Sync.SyncAdapter.ApplyChangesAsync (System.Guid localScopeId, System.Guid senderScopeId, Dotmim.Sync.SyncTable changesTable, System.Int64 lastTimestamp, System.Collections.Generic.List`1[T] conflicts, System.Data.Common.DbConnection connection, System.Data.Common.DbTransaction transaction) [0x003af] in <cc96665cb56c4641847da5ce877d6522>:0 
  at Dotmim.Sync.CoreProvider.ApplyChangesInternalAsync (Dotmim.Sync.SyncTable schemaTable, Dotmim.Sync.SyncContext context, Dotmim.Sync.MessageApplyChanges message, System.Data.Common.DbConnection connection, System.Data.Common.DbTransaction transaction, System.Data.DataRowState applyType, Dotmim.Sync.DatabaseChangesApplied changesApplied, System.Threading.CancellationToken cancellationToken, System.IProgress`1[T] progress) [0x004f5] in <cc96665cb56c4641847da5ce877d6522>:0 
  at Dotmim.Sync.CoreProvider.ApplyChangesAsync (Dotmim.Sync.SyncContext context, Dotmim.Sync.MessageApplyChanges message, System.Data.Common.DbConnection connection, System.Data.Common.DbTransaction transaction, System.Threading.CancellationToken cancellationToken, System.IProgress`1[T] progress) [0x00449] in <cc96665cb56c4641847da5ce877d6522>:0 
  at Dotmim.Sync.LocalOrchestrator.ApplyChangesAsync (Dotmim.Sync.ScopeInfo scope, Dotmim.Sync.SyncSet schema, Dotmim.Sync.Batch.BatchInfo serverBatchInfo, System.Int64 clientTimestamp, System.Int64 remoteClientTimestamp, Dotmim.Sync.Enumerations.ConflictResolutionPolicy policy, System.Threading.CancellationToken cancellationToken, System.IProgress`1[T] progress) [0x00364] in <cc96665cb56c4641847da5ce877d6522>:0 
   --- End of inner exception stack trace ---
  at Dotmim.Sync.BaseOrchestrator.RaiseError (System.Exception exception) [0x00046] in <cc96665cb56c4641847da5ce877d6522>:0 
  at Dotmim.Sync.LocalOrchestrator.ApplyChangesAsync (Dotmim.Sync.ScopeInfo scope, Dotmim.Sync.SyncSet schema, Dotmim.Sync.Batch.BatchInfo serverBatchInfo, System.Int64 clientTimestamp, System.Int64 remoteClientTimestamp, Dotmim.Sync.Enumerations.ConflictResolutionPolicy policy, System.Threading.CancellationToken cancellationToken, System.IProgress`1[T] progress) [0x00799] in <cc96665cb56c4641847da5ce877d6522>:0 
  at Dotmim.Sync.SyncAgent.SynchronizeAsync (Dotmim.Sync.Enumerations.SyncType syncType, System.Threading.CancellationToken cancellationToken, System.IProgress`1[T] progress) [0x00d1c] in <cc96665cb56c4641847da5ce877d6522>:0 

Thanks and Regards
Gireesh M

**Comment by mangireesh** [comment]
The error is solved by updating to 0.5.7.

**Comment by vunder** [comment]
@mangireesh I also had a problem with sync when `SyncId` column is not a primary key in table. This was happening while additional manipulation with sync. Ee.g. when syncing database that was not synced before and has some unique data, that are not present on the server. For this case I had addiional code
```csharp
                var agent = new SyncAgent(clientProvider, serverOrchestrator, new SyncOptions() { UseVerboseErrors = true }, syncSetup);
                var scope = await agent.LocalOrchestrator.GetClientScopeAsync(cts.Token);
                var s1 = await agent.SynchronizeAsync();
                if (scope.IsNewScope)
                {
                    // Get local schema
                    var schema = await agent.LocalOrchestrator.GetSchemaAsync(cts.Token, progress);

                    // Resetting primary keys to force to use SyncId as primary key
                    foreach (var t in schema.Tables)
                    {
                        t.PrimaryKeys.Clear();
                        t.PrimaryKeys.Add("SyncId");
                    }
                    // Update all rows on the client side that does not have any tracking rows
                    await agent.LocalOrchestrator.UpdateUntrackedRowsAsync(schema, cts.Token, progress);

                    s1 = await agent.SynchronizeAsync();
                }
```

I had to change list of primary keys in server schema, since they had information that is not matches to clients datbase schema

**Comment by mangireesh** [comment]
@vunder  Sorry for late reply. My problem was, I have a primary key and a identity column in my server table. Now when I do the sync operation, the table got synced, the identity column is got null in client. 
But when I make the identity column as Primary key then identity column will be populated but with different values as compared to server schema.

**Comment by Mimetis** [comment]
@mangireesh 
Can you share your table schema ?
I will make some tests on it

**Comment by sisaacks** [comment]
I am having the same type of issue, I am getting an error that Dotmim.Sync.Web.Client.HttpSyncWebException: Invalid column name 'Model' and also Dotmim.Sync.Web.Client.HttpSyncWebException: Invalid object name 'Vehicle_tracking

**Comment by Vextamour** [comment]
> I am having the same type of issue, I am getting an error that Dotmim.Sync.Web.Client.HttpSyncWebException: Invalid column name 'Model' and also Dotmim.Sync.Web.Client.HttpSyncWebException: Invalid object name 'Vehicle_tracking

Did you make sure you Provisioned the Azure SQL Server? This will create all the tracking tables for you in your Azure Cloud SQL Server. Sample code below but it's also in documentation. This goes in your WebAPI as end points. Then you can call it like this (you'll need to change URI to your pathway obviously): https://localhost:44368/api/DBMaintenance/ProvisionServerDatabase

Documentation: https://dotmimsync.readthedocs.io/Provision.html#provision-deprovision

Deprovision server if needed:
```
        //https://dotmimsync.readthedocs.io/Provision.html#deprovision
        [Route("[action]")]
        [HttpGet]
        //[Authorize]
        public async Task<string> DeprovisionServerDatabase()
        {            
            var serverDBConnectionString = Configuration.GetConnectionString("EndCapDbConnectionString");
            // Create server provider
            var serverProvider = new SqlSyncProvider(serverDBConnectionString);

            // Create standard Setup and Options
            var setup = new SyncSetup(Configuration["DBServerSyncTables:TableList"].Split(',').ToArray());
            var options = new SyncOptions { ConflictResolutionPolicy = Dotmim.Sync.Enumerations.ConflictResolutionPolicy.ServerWins };

            // Create a server orchestrator used to Deprovision everything on the server side
            var remoteOrchestrator = new RemoteOrchestrator(serverProvider, options, setup);

            // Deprovision everything
            await remoteOrchestrator.DeprovisionAsync();

            return "Server database successfully deprovisioned";
        }
```

Provision server (This is what you're needing):
```
        //https://dotmimsync.readthedocs.io/Provision.html#provision
        [Route("[action]")]
        [HttpGet]
        //[Authorize]
        public async Task<string> ProvisionServerDatabase()
        {
            var serverDBConnectionString = Configuration.GetConnectionString("EndCapDbConnectionString");
            var serverProvider = new SqlSyncProvider(serverDBConnectionString);

            // Create standard Setup and Options
            var setup = new SyncSetup(Configuration["DBServerSyncTables:TableList"].Split(',').ToArray());
            var options = new SyncOptions { ConflictResolutionPolicy = Dotmim.Sync.Enumerations.ConflictResolutionPolicy.ServerWins };

            // -----------------------------------------------------------------
            // Server side
            // -----------------------------------------------------------------

            // This method is useful if you want to provision by yourself the server database
            // You will need to :
            // - Create a remote orchestrator with the correct setup to create
            // - Provision everything

            // Create a server orchestrator used to Deprovision and Provision only table Address
            var remoteOrchestrator = new RemoteOrchestrator(serverProvider, options, setup);

            // Provision everything needed (sp, triggers, tracking tables)
            // Internally provision will fectch the schema a will return it to the caller.
            var newSchema = await remoteOrchestrator.ProvisionAsync();

            return "Server database successfully provisioned";
        }
```

**Comment by sisaacks** [comment]
I used entity framework to generate my schema in the table.  The table
schema is the same as the sqlite schema in the local database
This is getting increasingly confusing I keep getting different answers
from people

i have an Xamarin CrossPlatform Application.
I have a local SQLite Database I use
I have a Azure Database and web api.  I used entity framework to generate
the schema.
I cannot sync the two databases, I keep getting random different
errors...columns that I do not have in either database.
Does ANYONE have an example of going from Sqlite to SqlServer????

**Comment by Vextamour** [comment]
Sis, I provided all my code that I'm utilizing to have the databases sync above in prior responses from the WebAPI to the local application. My environment is .NET 6 WebAPI hosted in Azure with Azure SQL Server. I have an application in .NET Core that runs locally on their machine. When they run the project it creates the entire SQLite database, tables and syncs the data from the Azure SQL Server into the SQLite database for their application to read from. I also provided that code as well in the response. Scroll back up and look at my WebAPI code and application code to see if that helps you. I've provided working examples for both the API and Application side. I'm not quite sure how else I can help.

**Comment by sisaacks** [comment]
So I ended up deleting everything on Azure, re-adding everything and POOF now it works.  Something got mucked up.  I do have another question for you.  If I need to add a column to a table "Migrate" I am a little confused how to do that.  My app is for Mobile and its not a Console.  

When I add a new value to a model, on the client side, sqlite automatically updates the table.  So I am assuming I only need to update on the server side table.  But what I am confused about, is how do I call to the web api to do a migration?  I went to the github migration example but there is no explanation.
